### PR TITLE
Qtransform function

### PR DESCRIPTION
**Pull request by hagabbar**

### CODE REVIEW

**Comment by ahnitz** [comment]
@hagabbar Can you please rebase from master?

**Comment by hagabbar** [comment]
Rebased.

**Comment by ahnitz** [comment]
@hagabbar The function can be sped up *a lot*, but I don't see any style issues, so I'm OK with moving forward with this. We can handle performance issues in focused patches later on as necessary.

File: pycbc/filter/qtransform.py [hunk 1]
```diff
@@ -35,10 +35,72 @@
 from math import pi
 import numpy as np
 from pycbc.strain  import next_power_of_2
+from pycbc.types.timeseries import FrequencySeries, TimeSeries
+from numpy import fft as npfft
 
 __author__ = 'Hunter Gabbard <hunter.gabbard@ligo.org>'
 __credits__ = 'Duncan Macleod <duncan.macleod@ligo.org>'
 
+def qtransform(data, Q, f0, sampling, normalized):
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: normalized should be a keyword argument, no? And set to either False or True by default.

**Comment by hagabbar** [comment]
Will do. Any suggestions for further optimisation are always welcome!

File: pycbc/filter/qtransform.py (hunk 1 above)
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Ok, one other issue. You don't need sampling as a argument at all, I think. See inline how to simplify.

File: pycbc/filter/qtransform.py [hunk 2]
```diff
@@ -35,10 +35,72 @@
...
+    energy: '~pycbc.types.aligned.ArrayWithAligned'
+        A 'TimeSeries' of the complex energy from the Q-transform of 
+        this tile against the data.
+    """
+
+    # q-transform data for each (Q, frequency) tile
+
+    # initialize parameters
+    qprime = Q / 11**(1/2.)
+    dur = int(len(data)) / sampling
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Duration is just data.duration

File: pycbc/filter/qtransform.py [hunk 3]
```diff
@@ -35,10 +35,72 @@
...
+    fseries = TimeSeries.to_frequencyseries(data)
+
+    # window fft
+    window_size = 2 * int(f0 / qprime * dur) + 1
+
+    # get indices
+    indices = _get_indices(dur)
+
+    # apply window to fft
+    windowed = fseries[get_data_indices(dur, f0, indices)] * get_window(dur, indices, f0, qprime, Q, sampling)
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: 'sampling' can be replaced with fseries.delta_f

File: pycbc/filter/qtransform.py [hunk 4]
```diff
@@ -35,10 +35,72 @@
...
+
+    # pad data, move negative frequencies to the end, and IFFT
+    padded = np.pad(windowed, padding(window_size, dur, f0, Q), mode='constant')
+    wenergy = npfft.ifftshift(padded)
+
+    # return a 'TimeSeries'
+    wenergy = FrequencySeries(wenergy, delta_f=sampling)
```
Comment (ahnitz) [inline pycbc/filter/qtransform.py]: 'sampling' can be replaced with fseries.delta_f

File: pycbc/filter/qtransform.py (hunk 1 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: @ahnitz Would I define what normalized is set to be default within or outside the function?

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Just change to.

def qtransform(data, Q, f0, normalized=True)

Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Oh gotcha, I was going to do something with kwargs, but that makes more sense.

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: So, what I wrote makes it a keyword argument. I think you are thinking of functions that accept *arbitrary* keyword arguments as follows. We don't want to do that.

def f(**kwds)

File: pycbc/filter/qtransform.py (hunk 2 above)
Comment (hagabbar) [inline pycbc/filter/qtransform.py]: Because pycbc TimeSeries has that attribute?

Comment (ahnitz) [inline pycbc/filter/qtransform.py]: Yes.

**Pull request merged**